src/librustc_mir/interpret/eval_context.rs

```diff
@@ -756,6 +756,7 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
     pub(super) fn const_eval(
         &self,
         gid: GlobalId<'tcx>,
+        ty: Ty<'tcx>,
```

Comment:
So with this, the `gid` is not sufficient to determine the *type* of the constant it identifies? That seems somewhat odd to me, given that it is enough to determine the *value*.

Reply:
This is pretty much down to lifetime parametricity/polymorphism.
Basically the entire type is determinable *except* lifetimes.

And because lifetimes shouldn't affect const-evaluation, and to maximize caching, we want to erase them, but we also don't want the lifetime-erased type to leak back into e.g. typeck.

Reply:
This makes me think it would be nice to provide a "lifetime substitution" instead of the full type (and compute the non-lifetime parts from `GlobalId`), but that might need a lot more infrastructure than this PR (in general we might want lifetimes slots to be placeholders for e.g. MIR borrowck to work with).

cc @nikomatsakis (who I'm sure has had this idea before)

Reply:
> Basically the entire type is determinable except lifetimes.

Miri doesn't care about lifetimes though.

I am just worried about the additional failure mode this introduces -- generally in Miri we track the type of values so when someone uses e.g. an i32 as i64 we get an ICE (and this has caught quite a few bugs). But here we just believe the type the caller tells us.

Reply:
> Miri doesn't care about lifetimes though.

*The caller* does. Or is `const_eval` an internal API? (It's sometimes hard to tell on GitHub).

> But here we just believe the type the caller tells us.

If this is not internal, and this matters, we could assert that the type is identical after erasing the one that might have lifetimes.

If it *is* internal, then yeah, using whatever lifetime-erased type can be computed from `GlobalId` would be fine.

Reply:
> The caller does. Or is const_eval an internal API? (It's sometimes hard to tell on GitHub).

It's `pub(super)`, so only accessible inside the interpreter. Yes, it is internal.

Reply:
> It's `pub(super)`, so only accessible inside the interpreter.

Ahh, sorry, sometimes I confuse that for "may be used in the implementation of a query", but that's really `pub(crate)` because the CTFE queries are in the `rustc_mir::const_eval` module, right?

Reply:
Yes, this is mir interp internal. Rustc never calls this manually

Reply:
Indeed, that is exactly why I asked for the query impls to be moved out of `rustc_mir::interp` way back when.

Reply:
The `gid` should be sufficient to determine the type, at least a type that is good enough for miri, though getting the type of a promoted globals seems like it would be a little bit difficult. I decided as this is an internal API and as both call sites has the type available, it would just be easier to pass it through.

Reply:
One call site doesn't *really* have the type available... the one where you used `dest.layout.ty`, previously this was *checking* that the type of the constant was the same as the type of the place it was put into. Now there is no check any more, you are just assuming this to be the case.

Reply:
I will change that call site to actually use the proper type, which as it's for intrinsics the type of them are known.

Reply:
@skinny121 did you ever get around to actually do this? Doesn't look like it, but I might have missed it.